src/tir/ir/specialize.cc

```diff
@@ -0,0 +1,326 @@
+        const String& key = kv.first;
+        const ObjectRef& value = kv.second;
+        if (value->IsInstance<PrimExprNode>()) {
+          attr_dict.Set(key, Substitute(Downcast<PrimExpr>(value), var_map));
+        }
+      }
+    }
+    return f;
+  }
+
+ private:
+  Stmt VisitStmt_(const BlockNode* op) final {
+    Array<Buffer> alloc_buffers = MutateArray(
+        op->alloc_buffers,
+        std::bind(&PrimFuncSpecializer::MutateAllocBuffer, this, std::placeholders::_1));
+    Array<BufferRegion> reads = MutateArray(
+        op->reads,
+        std::bind(&PrimFuncSpecializer::MutateBufferRegion, this, std::placeholders::_1));
+    Array<BufferRegion> writes = MutateArray(
+        op->writes,
+        std::bind(&PrimFuncSpecializer::MutateBufferRegion, this, std::placeholders::_1));
+    Array<IterVar> block_vars = MutateArray(
+        op->iter_vars, std::bind(&PrimFuncSpecializer::MutateIterVar, this, std::placeholders::_1));
+    Optional<Stmt> init = NullOpt;
+    if (op->init.defined()) {
+      init = VisitStmt(op->init.value());
+    }
+    Stmt body = VisitStmt(op->body);
+
+    if (alloc_buffers.same_as(op->alloc_buffers) && reads.same_as(op->reads) &&
+        writes.same_as(op->writes) && block_vars.same_as(op->iter_vars) && body.same_as(op->body) &&
+        init.same_as(op->init)) {
+      return GetRef<Block>(op);
+    } else {
+      ObjectPtr<BlockNode> n = CopyOnWrite(op);
+      n->alloc_buffers = std::move(alloc_buffers);
+      n->reads = std::move(reads);
+      n->writes = std::move(writes);
+      n->iter_vars = std::move(block_vars);
+      n->body = std::move(body);
+      n->init = std::move(init);
+      return Stmt(n);
+    }
+  }
+
+  Stmt VisitStmt_(const BufferStoreNode* op) final {
+    auto it = buffer_map_.find(op->buffer);
+    if (it == buffer_map_.end()) {
+      return GetRef<BufferStore>(op);
+    }
+
+    PrimExpr value = VisitExpr(op->value);
+    Array<PrimExpr> indices =
+        MutateArray(op->indices, [this](const PrimExpr& e) { return this->VisitExpr(e); });
+
+    auto n = CopyOnWrite(op);
+    n->buffer = it->second;
+    n->value = std::move(value);
+    n->indices = std::move(indices);
+    return Stmt(n);
+  }
+
+  PrimExpr VisitExpr_(const BufferLoadNode* op) final {
+    auto it = buffer_map_.find(op->buffer);
+    if (it == buffer_map_.end()) {
+      return GetRef<BufferLoad>(op);
+    }
+
+    Array<PrimExpr> indices =
+        MutateArray(op->indices, [this](const PrimExpr& e) { return this->VisitExpr(e); });
+
+    auto n = CopyOnWrite(op);
+    n->buffer = it->second;
+    n->indices = std::move(indices);
+    return PrimExpr(n);
+  }
+
+  PrimExpr VisitExpr_(const VarNode* op) final {
+    auto it = var_map_.find(GetRef<Var>(op));
+    if (it == var_map_.end()) {
+      return GetRef<PrimExpr>(op);
+    } else {
+      return it->second;
+    }
+  }
+
+ private:
+  Buffer MutateBuffer(Buffer buffer) const {
+    BufferNode* buffer_ptr = buffer.CopyOnWrite();
```

Comment:
The following invariance must hold: if nothing is remapped, we return the same buffer. Right now this does not hold.

- We need to check the new shape, strides